native/wasmex/src/component_type_conversion.rs

```diff
@@ -81,53 +95,146 @@ pub fn term_to_val(param_term: &Term, param_type: &Type) -> Result<Val, Error> {
                 .map(|(key_term, val)| (term_to_field_name(key_term), val))
                 .collect::<Vec<(String, &Term)>>();
             for field in record.fields() {
-                let field_term_option = terms.iter().find(|(k, _)| k == field.name);
+                let field_term_option = terms
+                    .iter()
+                    .find(|(field_name, _)| field_name == field.name);
                 if let Some((_, field_term)) = field_term_option {
                     let field_value = term_to_val(field_term, &field.ty)?;
                     kv.push((field.name.to_string(), field_value))
                 }
             }
+            if kv.len() != record.fields().len() {
+                let missing_fields = record
+                    .fields()
+                    .filter(|field| !terms.iter().any(|(name, _)| name == field.name))
+                    .collect::<Vec<_>>();
+                return Err(Error::Term(Box::new(format!(
+                    "Expected {} fields, got {} - missing fields: {}",
+                    record.fields().len(),
+                    kv.len(),
+                    missing_fields
+                        .iter()
+                        .map(|field| field.name)
+                        .collect::<Vec<_>>()
+                        .join(", ")
+                ))));
+            }
             Ok(Val::Record(kv))
         }
-        (TermType::Atom, Type::Option(option_type)) => {
-            let the_atom = param_term.atom_to_string()?;
-            if the_atom == "nil" {
+        (TermType::Atom, Type::Option(_option_type)) => {
+            let none_atom = param_term.atom_to_string()?;
+            if none_atom == "none" {
                 Ok(Val::Option(None))
             } else {
-                let converted_val = term_to_val(param_term, &option_type.ty())?;
-                Ok(Val::Option(Some(Box::new(converted_val))))
+                Err(Error::Term(Box::new(format!(
+                    "Invalid atom: {}, expected ':none' or '{{:some, term}}' tuple",
+                    none_atom
+                ))))
+            }
+        }
```

Comment:
here we need to add a test whether `option_type` is `None` - only then are we allowed to use the `:none` case. 

-> Can be done in a later PR to untangle this change from this already gigantic PR